### PR TITLE
plugins/schemastore: init

### LINKED ISSUE #1443: [PLUGIN REQUEST] `SchemaStore.nvim`
| Field    | Description |
|----------|-------------|
| Plugin   | `SchemaStore.nvim`  | 
| Homepage | [SchemaStore.nvim](https://github.com/b0o/SchemaStore.nvim)     | 
| Nixpkgs  | [vimPlugins.SchemaStore-nvim](https://search.nixos.org/packages?channel=unstable&show=vimPlugins.SchemaStore-nvim) | 

## Extra Information

From repo's README:
> A Neovim plugin that provides the [SchemaStore](https://github.com/SchemaStore/schemastore) catalog for use with [jsonls](https://github.com/hrsh7th/vscode-langservers-extracted) and [yamlls](https://github.com/redhat-developer/yaml-language-server).

### PR DESCRIPTION
**Pull request by GaetanLepage**
Add support for the [SchemaStore.nvim](https://github.com/b0o/SchemaStore.nvim/tree/main) plugin, providing JSON schemas for Neovim.

Fixes #1443

### CODE REVIEW

**Comment by GaetanLepage** [comment]
CI now passes. This PR is ready for merging.

File: plugins/lsp/language-servers/default.nix
Comment (traxys) [inline plugins/lsp/language-servers/default.nix:1]: Those changes seem unrelated ?

File: plugins/lsp/schemastore.nix
Comment (traxys) [inline plugins/lsp/schemastore.nix:137]: Why is this commented ?

File: tests/test-sources/plugins/lsp/yamlls.nix
Comment (traxys) [inline tests/test-sources/plugins/lsp/yamlls.nix:1]: Unrelated?

File: tests/test-sources/plugins/lsp/jsonls.nix
Comment (traxys) [inline tests/test-sources/plugins/lsp/jsonls.nix:1]: Unrelated?

**Review by traxys — COMMENTED** [review]
I'm pretty sure this is not the correct way to package this plugin. This plugin is a bit strange in that it provides helper functions for jsonls/yamls.
I think the most idiomatic way to use this plugin would be to provide the _same_ helper functions in some way, but that seems a bit against the module system.

File: plugins/lsp/schemastore.nix
Comment (MattSturgeon) [reply plugins/lsp/schemastore.nix:137]: Without this being uncommented, this module has no real effect on `init.lua`

File: plugins/lsp/schemastore.nix
Comment (MattSturgeon) [inline plugins/lsp/schemastore.nix:92]: Seems odd to have `replace` and `extra` examples without `select` and `ignore`. Maybe they could be `yaml` examples? :thinking:

File: plugins/lsp/schemastore.nix
Comment (MattSturgeon) [inline plugins/lsp/schemastore.nix:42]: Could link to [the catalogue](https://github.com/SchemaStore/schemastore/blob/master/src/api/json/catalog.json) for reference

File: plugins/lsp/language-servers/default.nix
Comment (MattSturgeon) [reply plugins/lsp/language-servers/default.nix:1]: Looks like adding `settings` to `yamlls` and `jsonls` was required in order to [set the schema configs](https://github.com/nix-community/nixvim/pull/1477/files#diff-ac9c057b00cddb12c87df1b5a245e4188340520b6507fba4602c3adc5e2298edR121-R144).

I've extracted them into #1563 in case we'd like to review seperately or merge early.

File: plugins/lsp/schemastore.nix
Comment (MattSturgeon) [inline plugins/lsp/schemastore.nix:119]: Should this have a warning/assertion checking that the respective LSP is also enabled when _this_ is enabled?

Comment (MattSturgeon) [reply plugins/lsp/schemastore.nix:123]:
```suggestion
  extraConfig = cfg: {
    warnings =
      (optional (!(cfg.json.enable || cfg.yaml.enable)) ''
        NixVim(plugins.schemastore): you have enabled the plugin, but neither `json` or `yaml` schemas are enabled.
      '')
      ++ (optional (!(cfg.json.enable -> config.plugins.lsp.servers.jsonls.enable)) ''
        NixVim(plugins.schemastore): you have enabled `json` schemas, but `plugins.lsp.servers.jsonls` is not enabled.
      '')
      ++ (optional (!(cfg.yaml.enable -> config.plugins.lsp.servers.yamlls.enable)) ''
        NixVim(plugins.schemastore): you have enabled `yaml` schemas, but `plugins.lsp.servers.yamlls` is not enabled.
      '');

```

File: tests/test-sources/plugins/lsp/schemastore.nix
Comment (MattSturgeon) [inline tests/test-sources/plugins/lsp/schemastore.nix:48]: Would be nice to have additional `withJson` and `withYaml` tests to verify it's still valid when only one is in use.

File: plugins/lsp/schemastore.nix
Comment (MattSturgeon) [inline plugins/lsp/schemastore.nix:118]: Is there a reason the json settings option has an example but yaml does not? Upstream docs lacking an example?

File: plugins/lsp/schemastore.nix
Comment (MattSturgeon) [inline plugins/lsp/schemastore.nix:20]: Should this be freeformType as it corresponds to plugin settings?

File: plugins/lsp/schemastore.nix
Comment (MattSturgeon) [inline plugins/lsp/schemastore.nix:25]: This needs `maybeRaw`, so either:

```suggestion
              name = helpers.mkNullOrStr "The name of the schema.";
```

Or (if you don't want `nullOr`):

```suggestion
              name = mkOption {
                type = with helpers.nixvimTypes; maybeRaw str;
                description = "The name of the schema.";
              };
```

File: plugins/lsp/schemastore.nix
Comment (MattSturgeon) [inline plugins/lsp/schemastore.nix:29]: Raw types:

```suggestion
              fileMatch = helpers.mkNullOrOption (with helpers.nixvimTypes; maybeRaw (either str (listOf (maybeRaw str)))) ''
```

File: plugins/lsp/schemastore.nix
Comment (MattSturgeon) [inline plugins/lsp/schemastore.nix:36]: Same as feedback as `name`: use `mkNullOrStr` or `maybeRaw`

File: plugins/lsp/schemastore.nix
Comment (MattSturgeon) [inline plugins/lsp/schemastore.nix:41]: Update to new non-string defaults:

```suggestion
          select = helpers.defaultNullOpts.mkListOf types.str [ ] ''
```

File: plugins/lsp/schemastore.nix
Comment (MattSturgeon) [inline plugins/lsp/schemastore.nix:50]: Same
```suggestion
          ignore = helpers.defaultNullOpts.mkListOf types.str [ ] ''
```

File: plugins/lsp/schemastore.nix
Comment (MattSturgeon) [inline plugins/lsp/schemastore.nix:57]: Same
```suggestion
          replace = helpers.defaultNullOpts.mkAttrsOf schemaEntry { } ''
```

File: plugins/lsp/schemastore.nix
Comment (MattSturgeon) [inline plugins/lsp/schemastore.nix:60]: New paragraph:
```suggestion
            An attrs of elements representing schemas to replace with a custom schema.

            The string key is the name of the schema to replace, the table value is the schema definition.
            If a schema with the given name isn't found, the custom schema will not be returned.
```

File: plugins/lsp/schemastore.nix
Comment (MattSturgeon) [inline plugins/lsp/schemastore.nix:63]: Same
```suggestion
          extra = helpers.defaultNullOpts.mkListOf schemaEntry [ ] ''
```

**Review by MattSturgeon — CHANGES_REQUESTED** [review]
I've given this another pass, as it'd be good to have merged.

Most feedback relates to supporting the raw type.

@traxys do the warning's I've suggested resolve your concerns with the way the plugin is integrated, or do you still think we should be taking a different approach?\
If not, is that important enough to block this PR, or should we merge and consider improvements later?

File: plugins/lsp/schemastore.nix
Comment (GaetanLepage) [reply plugins/lsp/schemastore.nix]: I use `mkOption` here to make this option mandatory.
I have changed the type to `maybeRaw str` though.

File: plugins/lsp/schemastore.nix
Comment (GaetanLepage) [reply plugins/lsp/schemastore.nix:20]: I have added `freeformType`.

Comment (MattSturgeon) [reply plugins/lsp/schemastore.nix:20]: I'm still conflicted with submodules; on the one hand we're specifying some options can't be null and on the other hand we're saying it's a freeform type :joy: 

I guess freeform only really affects undeclared options?

Comment (GaetanLepage) [reply plugins/lsp/schemastore.nix:20]: Yes, exactly. If you explicitly declare an option, it will be constrained in terms of type.
The `freeformType` allows you to provide values for arbitrary undeclared options.
They are not at all contradictory.

File: plugins/lsp/schemastore.nix
Comment (GaetanLepage) [reply plugins/lsp/schemastore.nix:122]: Yes, lack of example upstream. Also, I think it is fine not to put examples absolutely everywhere.

File: tests/test-sources/plugins/lsp/schemastore.nix
Comment (MattSturgeon) [reply tests/test-sources/plugins/lsp/schemastore.nix:48]: E.g.

```nix
withJson = {
  plugins = {
    lsp = {
      enable = true;
      servers.jsonls.enable = true;
    };

    schemastore = {
      enable = true;

      json = {
        enable = true;
        settings = { };
      };
    };
  };
};

withYaml = {
  plugins = {
    lsp = {
      enable = true;
      servers.yamlls.enable = true;
    };

    schemastore = {
      enable = true;

      yaml = {
        enable = true;
        settings = { };
      };
    };
```

Comment (GaetanLepage) [reply tests/test-sources/plugins/lsp/schemastore.nix:48]: True, done.

File: plugins/lsp/schemastore.nix
Comment (GaetanLepage) [reply plugins/lsp/schemastore.nix:107]: I have not found existing examples online. I don't know if we should seek exhaustive examples.

File: plugins/lsp/schemastore.nix
Comment (MattSturgeon) [inline plugins/lsp/schemastore.nix:20]: Personally, I'd move `schemaEntry` to the `extraOptions` let-block (alongside `schemaOpts`). I find that more readable, rather than using doubly nested lets, but personal preference.

Comment (GaetanLepage) [reply plugins/lsp/schemastore.nix]: Fair, done.

**Review by MattSturgeon — APPROVED** [review]
Looks good! Thanks for being patient with me.

Let's give @traxys a couple hours to comment before merging, since he had concerns before.

**Review by traxys — APPROVED** [review]

**Pull request merged**
